File: mini1/src/airnow-detailed/parallel.py
from concurrent.futures import ThreadPoolExecutor
import os
import time

# Define a class to represent each data entry
class DataEntry:
    def __init__(self, latitude, longitude, site_name, parameter, concentration):
        self.latitude = latitude
        self.longitude = longitude
        self.site_name = site_name
        self.parameter = parameter
        self.concentration = float(concentration)

def read_data(thread_count=1):
    # Define the directory path
    directory_path = '../../data/AirNow Fires/data/'
    # List to store all file names
    file_names = []
    # Walk through the directory to get all .csv file names
    for root, dirs, files in os.walk(directory_path):
        for file in files:
            if file.endswith('.csv'):
                file_names.append(os.path.join(root, file))

    print('Number of files: ', len(file_names))

    # List to store all data entries (objects)
    data_entries = []

    # Function to process each chunk (file)
    def process_chunk(file):
        with open(file, 'r') as f:
            lines = f.readlines()
            print('Processing file: ', file)
            for line in lines[1:]:  # Skip header line
                line = line.strip().replace('"', '').split(',')
                if float(line[7]) > 0:  # Ensure concentration is positive
                    # Create a data entry object and add to the list
                    entry = DataEntry(
                        latitude=line[0],
                        longitude=line[1],
                        site_name=line[9],
                        parameter=line[3],
                        concentration=line[7]
                    )
                    data_entries.append(entry)

    # Process each file concurrently using ThreadPoolExecutor
    with ThreadPoolExecutor(max_workers=thread_count) as executor:
        executor.map(lambda file: process_chunk(file), file_names)

    # Function to calculate the max concentrations
    def calculate_max(entries):
        aggregated_data = {}
        for entry in entries:
            key = f"{entry.site_name} {entry.latitude} {entry.longitude}"
            if key not in aggregated_data:
                aggregated_data[key] = {}
            if entry.parameter not in aggregated_data[key]:
                aggregated_data[key][entry.parameter] = 0
            aggregated_data[key][entry.parameter] += entry.concentration
        
        # Now find the maximum value for each key
        max_data = {}
        for key in aggregated_data:
            max_data[key] = max(aggregated_data[key], key=aggregated_data[key].get)
        return max_data

    # Split data into batches for processing
    def split_into_batches(data, batch_size):
        for i in range(0, len(data), batch_size):
            yield data[i:i + batch_size]

    data_batches = list(split_into_batches(data_entries, len(data_entries) // thread_count + 1))

    # Calculate max for each batch using threads
    max_data_results = []
    with ThreadPoolExecutor(max_workers=thread_count) as executor:
        max_data_results = list(executor.map(calculate_max, data_batches))

    # Combine all max_data results from different batches
    combined_max_data = {}
    for max_data in max_data_results:
        combined_max_data.update(max_data)

    print('Number of unique site-parameter combinations with max data: ', len(combined_max_data))

start = time.time()
read_data(6)
print('Time taken: ', time.time() - start)

File: mini1/src/airnow-detailed/parallel.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <unordered_map>
#include <vector>
#include <filesystem>
#include <omp.h>
#include <chrono>

using namespace std;

struct DataPoint {
    string key;
    string parameter;
    double concentration;

    DataPoint(const string& k, const string& p, double c)
        : key(k), parameter(p), concentration(c) {}
};

vector<string> get_file_names(const string& directory_path) {
    vector<string> csv_files;
    for (const auto& entry : filesystem::recursive_directory_iterator(directory_path)) {
        if (entry.is_regular_file() && entry.path().extension() == ".csv") {
            csv_files.push_back(entry.path().string());
        }
    }
    cout << "Total number of files: " << csv_files.size() << endl;
    return csv_files;
}

const int BUFFER_SIZE = 512 * 1024; // 512KB buffer

void process_buffer(char* buffer, size_t size, vector<DataPoint>& data_points) {
    string key, parameter;
    double concentration;
    string token;
    vector<string> tokens;
    bool insideQuotes = false;

    for (size_t i = 0; i < size; ++i) {
        char c = buffer[i];

        if (c == '"') {
            insideQuotes = !insideQuotes;
            if (!insideQuotes) {
                tokens.push_back(token);
                token.clear();
            }
            continue;
        }

        if (insideQuotes && c != ',') {
            token += c;
        }
        
        // Check for new line
        if (c == '\n') {
            // Ensure we have enough tokens to process
            if (tokens.size() > 9) {
                key = tokens[9] + " " + tokens[0] + " " + tokens[1];
                concentration = stod(tokens[7]);
                parameter = tokens[3];
                if (concentration > 0) {
                    #pragma omp critical
                    {
                        data_points.emplace_back(key, parameter, concentration);
                    }
                }
            }
            tokens.clear();
        }
    }
}

void process_file(const string& file_name, vector<DataPoint>& data_points) {
    ifstream file(file_name, ios::in | ios::binary);
    cout << "Processing file: " << file_name << endl;
    if (!file) {
        cerr << "Error opening file: " << file_name << endl;
        return;
    }

    char buffer[BUFFER_SIZE];

    while (file) {
        file.read(buffer, BUFFER_SIZE);
        streamsize count = file.gcount();

        if (count == 0) break;

        process_buffer(buffer, count, data_points);
    }

    file.close();
}

void calculate_max(const vector<DataPoint>& data_points, unordered_map<string, string>& max_data) {
    unordered_map<string, double> max_concentration;

    for (const auto& data_point : data_points) {
        const auto& key = data_point.key;
        if (data_point.concentration > max_concentration[key]) {
            max_concentration[key] = data_point.concentration;
            max_data[key] = data_point.parameter; // Store the corresponding parameter
        }
    }
}

int main() {
    ios::sync_with_stdio(false);

    string directory_path = "../data/AirNow Fires/data/";
    double start_time = omp_get_wtime();

    vector<string> file_names = get_file_names(directory_path);

    vector<DataPoint> data_points; // Store DataPoint objects
    unordered_map<string, string> max_data;

    int thread_count = 4;

    // Process files in parallel
    #pragma omp parallel for num_threads(thread_count)
    for (int i = 0; i < file_names.size(); i++) {
        vector<DataPoint> local_data_points; // Local storage for each thread
        process_file(file_names[i], local_data_points);

        // Merge results into the shared data_points vector
        #pragma omp critical
        {
            data_points.insert(data_points.end(), local_data_points.begin(), local_data_points.end());
        }
    }

    // Calculate maximums
    calculate_max(data_points, max_data);

    for (const auto& entry : max_data) {
        cout << entry.first << " -> " << entry.second << endl;
    }

    cout << "Total number of unique keys processed: " << max_data.size() << endl;

    double end_time = omp_get_wtime();
    cout << "Time taken: " << (end_time - start_time) << " seconds" << endl;

    return 0;
}
